quickwit-serve/src/test_utils/cluster_sandbox.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::time::Duration;
+
+use hyper::Uri;
+use itertools::Itertools;
+use quickwit_cluster::QuickwitService;
+use quickwit_common::new_coolid;
+use quickwit_common::uri::Uri as QuickwitUri;
+use quickwit_config::QuickwitConfig;
+use quickwit_metastore::{quickwit_metastore_uri_resolver, IndexMetadata};
+use quickwit_proto::tonic::transport::Endpoint;
+use quickwit_search::{create_search_service_client, SearchServiceClient};
+use rand::seq::SliceRandom;
+use tempfile::TempDir;
+
+use super::rest_client::QuickwitRestClient;
+use crate::serve_quickwit;
+
+/// Configuration of a node made of a [`QuickwitConfig`] and a
+/// set of services.
+#[derive(Clone)]
+pub struct NodeConfig {
+    pub quickwit_config: QuickwitConfig,
+    pub services: HashSet<QuickwitService>,
+}
+
+/// Creates a Cluster Test environment.
+///
+/// The goal is to start several nodes and use the gRPC or REST clients to
+/// test it.
+pub struct ClusterSandbox {
+    pub node_configs: Vec<NodeConfig>,
+    pub grpc_search_clients: HashMap<SocketAddr, SearchServiceClient>,
+    pub rest_client: QuickwitRestClient,
+    pub index_id_for_test: String,
+    _temp_dir: TempDir,
+}
+
+impl ClusterSandbox {
+    // Starts one node that runs all the services.
+    pub async fn start_standalone_node() -> anyhow::Result<Self> {
+        let temp_dir = tempfile::tempdir()?;
+        let services = HashSet::from_iter([QuickwitService::Searcher, QuickwitService::Indexer]);
+        let node_configs = build_node_configs(temp_dir.path().to_path_buf(), vec![services]);
+        // There is exactly one node.
+        let node_config = node_configs[0].clone();
+        let node_config_clone = node_config.clone();
+        // Creates an index before starting nodes as currently Quickwit does not support
+        // dynamic creation/deletion of indexes/sources.
+        let index_for_test = new_coolid("index-for-test");
```

Comment:
There's also `append_random_suffix` in `common`. I usually name the index after the test so that if there's an issue with the index, I quickly know which test created that index. For instance: `append_random_suffix("test-standalone-node")`.